File: v1/code/go/src/jvmgo/ch11/rtda/local_vars.go
package rtda

import "math"
import "jvmgo/ch11/rtda/heap"

type LocalVars []Slot

func newLocalVars(maxLocals uint) LocalVars {
	if maxLocals > 0 {
		return make([]Slot, maxLocals)
	}
	return nil
}

func (self LocalVars) SetInt(index uint, val int32) {
	self[index].num = val
}
func (self LocalVars) GetInt(index uint) int32 {
	return self[index].num
}

func (self LocalVars) SetFloat(index uint, val float32) {
	bits := math.Float32bits(val)
	self[index].num = int32(bits)
}
func (self LocalVars) GetFloat(index uint) float32 {
	bits := uint32(self[index].num)
	return math.Float32frombits(bits)
}

// long consumes two slots
func (self LocalVars) SetLong(index uint, val int64) {
	self[index].num = int32(val)
	self[index+1].num = int32(val >> 32)
}
func (self LocalVars) GetLong(index uint) int64 {
	low := uint32(self[index].num)
	high := uint32(self[index+1].num)
	return int64(high)<<32 | int64(low)
}

// double consumes two slots
func (self LocalVars) SetDouble(index uint, val float64) {
	bits := math.Float64bits(val)
	self.SetLong(index, int64(bits))
}
func (self LocalVars) GetDouble(index uint) float64 {
	bits := uint64(self.GetLong(index))
	return math.Float64frombits(bits)
}

func (self LocalVars) SetRef(index uint, ref *heap.Object) {
	self[index].ref = ref
}
func (self LocalVars) GetRef(index uint) *heap.Object {
	return self[index].ref
}

func (self LocalVars) SetSlot(index uint, slot Slot) {
	self[index] = slot
}

func (self LocalVars) GetThis() *heap.Object {
	return self.GetRef(0)
}
func (self LocalVars) GetBoolean(index uint) bool {
	return self.GetInt(index) == 1
}

File: v1/code/java/jvmgo_java/ch05/src/main/java/com/github/jvmgo/rtda/LocalVars.java
package com.github.jvmgo.rtda;

import lombok.Getter;
import lombok.Setter;
import lombok.ToString;

import java.lang.ref.Reference;

/**
 */
@Setter
@Getter
@ToString
public class LocalVars {

    private Slot[] localVars;

    public LocalVars(int maxLocals) {
        if (maxLocals > 0)
            localVars = new Slot[maxLocals];
        else
            System.exit(-1);
    }

    public void setInt(int index, int val) {
        localVars[index] = Util.setInt(val);
    }

    public int getInt(int index) {
        Slot slot = localVars[index];
        return Util.getInt(slot);
    }

    public void setFloat(int index, float val) {
        int a = Float.floatToIntBits(val);
        setInt(index, a);
    }

    public float getFloat(int index) {
        int a = getInt(index);
        return Float.intBitsToFloat(a);
    }

    public void setLong(int index, long val) {
        Slot[] slots = Util.setLong(val);
        localVars[index] = slots[0];
        localVars[index + 1] = slots[1];
    }

    public long getLong(int index) {
        Slot high = localVars[index];
        Slot low = localVars[index + 1];
        return Util.getLong(new Slot[]{high, low});
    }

    public void setDouble(int index, double val) {
        long v = Double.doubleToLongBits(val);
        setLong(index, v);
    }

    public double getDouble(int index) {
        long v = getLong(index);
        return Double.longBitsToDouble(v);
    }

    public void setRef(int index, Reference ref) {
        Slot slot = new Slot();
        slot.setRef(ref);
        localVars[index] = slot;
    }

    public Reference getRef(int index) {
        return localVars[index].getRef();
    }
}
